lib/ProductOpener/ProductSchemaChanges.pm

```diff
@@ -222,3 +235,83 @@ sub convert_schema_1001_to_1000_remove_ingredients_hierarchy_taxonomize_brands (
 	return;
 }
 
+=head2 1001 to 1002 - Refactor the images object
+
+The images object is restructured to separate uploaded images from selected images
+
+=cut
+
+sub convert_schema_1001_to_1002_refactor_images_object ($product_ref) {
+
+	if (exists $product_ref->{images}) {
+		my $uploaded_ref = {};
+		my $selected_ref = {};
+		foreach my $imgid (%{$product_ref->{images}}) {
+			# Uploaded images have an imageid which is a string containing an integer starting from 1
+			if ($imgid =~ /^\d+$/) {
+				$uploaded_ref->{$imgid} = $product_ref->{images}->{$imgid};
+			}
+			# Selected images have an imageid which is a string containing a word and a 2 letter language code (e.g. ingredients_fr)
+			elsif ($imgid =~ /^(\w+)_(\w\w)$/) {
+				my ($type, $lc) = ($1, $2);
+				# Put keys related to cropping / rotation / normalization inside a "generation" structure
+				my $image_ref = $product_ref->{images}->{$imgid};
+				my $new_image_ref = {
+					imgid => $image_ref->{imgid},
+					rev => $image_ref->{rev},
+					sizes => $image_ref->{sizes},
+				};
+				my $generation_ref = normalize_generation_ref($image_ref);
+				if (defined $generation_ref) {
+					$new_image_ref->{generation} = $generation_ref;
+				}
+				deep_set($selected_ref, $type, $lc, $new_image_ref);
+			}
+		}
+		$product_ref->{images} = {
+			uploaded => $uploaded_ref,
+			selected => $selected_ref,
+		};
+	}
+
+	return;
+}
+
+sub convert_schema_1002_to_1001_refactor_images_object ($product_ref) {
+
+	# We need to convert the images object back to the old format
+	if (exists $product_ref->{images}) {
+		my $images_ref = $product_ref->{images};
+		# Copy uploaded images
+		if (exists $images_ref->{uploaded}) {
+			# We need to copy the uploaded images back to the main images object
+			foreach my $imgid (keys %{$images_ref->{uploaded}}) {
+				$product_ref->{images}->{$imgid} = $images_ref->{uploaded}{$imgid};
+			}
+			delete $product_ref->{images}->{uploaded};
```

Comment:
nitpick: I would better have built a new hasmap, new_images_ref.

Can't it have side effect this way ?

Reply:
Sure, changing.